examples/async-sign/src/lib.rs

```diff
@@ -1,82 +1,81 @@
+}
 
-    hex::decode(hex)
-        .map(|bytes| ActorId::from_slice(&bytes).expect("Unable to create ActorId"))
-        .map_err(|_| ())
+gstd::metadata! {
+    title: "demo async sign",
+    init:
+        input: InputArgs,
 }
 
 #[no_mangle]
 pub unsafe extern "C" fn init() {
-    let input = String::from_utf8(msg::load_bytes()).expect("Invalid message: should be utf-8");
-    let dests: Vec<&str> = input.split(',').collect();
-    if dests.len() != 2 {
-        panic!("Invalid input, should be two IDs separated by comma");
-    }
-    SIGNATORY = hex_to_id(dests[0]).expect("INTIALIZATION FAILED: INVALID PROGRAM ID");
-    SIGNED_MESSAGE_PROGRAM = hex_to_id(dests[1]).expect("INTIALIZATION FAILED: INVALID PROGRAM ID");
+    let args: InputArgs = msg::load().expect("Failed to decode `InputArgs`");
+
+    DESTINATION = args.destination;
+    SIGNATORY = args.signatory;
 }
 
 #[gstd::async_main]
 async fn main() {
-    let message = String::from_utf8(msg::load_bytes()).expect("Invalid message: should be utf-8");
+    let message = msg::load_bytes();
     debug!("message = {:?}", message);
-    if message == "START" {
-        let encoded = SignRequest {
-            message: b"PING".to_vec(),
-        }
-        .encode();
-
-        let sign_response =
-            msg::send_bytes_and_wait_for_reply(unsafe { SIGNATORY }, &encoded, GAS_LIMIT, 0)
-                .await
-                .expect("Error in async message processing");
 
-        debug!("sign_response = {:?}", sign_response);
+    let request = SignRequest {
+        message: message.clone(),
+    };
 
-        let reply = msg::send_bytes_and_wait_for_reply(
-            unsafe { SIGNED_MESSAGE_PROGRAM },
-            &sign_response,
-            GAS_LIMIT,
-            0,
-        )
-        .await
-        .expect("Error in async message processing");
+    let sign_response: Result<SignResponse, _> =
+        msg::send_and_wait_for_reply(unsafe { SIGNATORY }, &request, GAS_LIMIT, 0).await;
 
-        debug!("reply = {:?}", reply);
+    let verified = sign_response
+        .ok()
+        .and_then(|r| {
+            // the same way as in verify.rs from subkey
+            let mut signature: Signature = Default::default();
+            if r.signature.len() == signature.0.len() {
+                signature.as_mut().copy_from_slice(&r.signature);
+                Some(signature)
+            } else {
+                None
+            }
+        })
+        .map(|signature| {
+            let pub_key = Public::unchecked_from(<[u8; 32]>::from(unsafe { SIGNATORY }));
 
-        let result = format!(
-            "{}{}",
-            FINISH_STRING,
-            str::from_utf8(&reply).expect("Failed to interpret `reply` as utf-8")
-        );
+            Sr25519Pair::verify(&signature, &message, &pub_key)
+        })
+        .unwrap_or(false);
 
-        msg::reply_bytes(result, GAS_LIMIT, 0);
+    if verified {
+        msg::send_bytes(unsafe { DESTINATION }, message, GAS_LIMIT, 0);
```

Comment:
```suggestion
        let SignRequest {
            message,
        } = request;
        msg::send_bytes(unsafe { DESTINATION }, message, GAS_LIMIT, 0);
```

Reply:
destructuring takes more lines so I decided to move the field out of SignRequest